core/src/main/scala/org/broadinstitute/dsde/rawls/coordination/DataSourceAccess.scala

```diff
@@ -0,0 +1,63 @@
+package org.broadinstitute.dsde.rawls.coordination
+
+import akka.actor.ActorRef
+import akka.pattern.ask
+import akka.util.Timeout
+import org.broadinstitute.dsde.rawls.dataaccess.SlickDataSource
+import org.broadinstitute.dsde.rawls.dataaccess.slick.{DataAccess, ReadWriteAction}
+import slick.jdbc.TransactionIsolation
+
+import scala.concurrent.Future
+import scala.concurrent.duration.{Deadline, FiniteDuration}
+import scala.reflect.ClassTag
+
+/**
+ * Queries data access, possibly running one future at a time to avoid deadlocks. WA-177
+ */
+trait DataSourceAccess {
+  val slickDataSource: SlickDataSource
+
+  def inTransaction[A: ClassTag](dataAccessFunction: DataAccess => ReadWriteAction[A],
+                                 isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead,
+                                ): Future[A]
+}
+
+/**
+ * Send all data access requests as concurrently as possible, even if deadlocks may occur.
+ */
+class UncoordinatedDataSourceAccess(override val slickDataSource: SlickDataSource) extends DataSourceAccess {
+
+  override def inTransaction[A: ClassTag](dataAccessFunction: DataAccess => ReadWriteAction[A],
+                                          isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead,
+                                         ): Future[A] = {
+    slickDataSource.inTransaction(dataAccessFunction, isolationLevel)
+  }
+}
+
+/**
+ * Queue up the request in an underlying actor's mailbox, that will run only one request at a time, primarily to avoid
+ * deadlocks. NOTE: Ideally the call to inTransaction should only contain IO requests, and should not spend significant
+ * CPU processing time, as other threads/actors will be waiting for the exclusive use of the shared actor.
+ */
+class CoordinatedDataSourceAccess(override val slickDataSource: SlickDataSource,
+                                  dataSourceActor: ActorRef,
+                                  starTimeout: FiniteDuration,
+                                  waitTimeout: FiniteDuration,
+                                  askTimeout: FiniteDuration,
+                                 ) extends DataSourceAccess {
+  override def inTransaction[A: ClassTag](dataAccessFunction: DataAccess => ReadWriteAction[A],
+                                          isolationLevel: TransactionIsolation): Future[A] = {
+    // How long to wait for the actor to get to and finish running the function
+    implicit val implicitTimeout: Timeout = Timeout(askTimeout)
+    // Throws a DeadlineTimeoutException the function is not started within startTimeout
+    // Throws a TimeoutException if the function does not run within the waitTimeout
+    // Throws an AskTimeoutException if we do not hear back either way before askTimeout
```

Comment:
Are these exceptions handled in the SubmissionMonitorActor? Or are they retried somehow?

Reply:
These are new Exceptions, however the SMA has an Akka supervisor that retries non-fatal errors.